admin/server/github.go

```diff
@@ -853,9 +972,294 @@ func (s *Server) fetchReposForInstallation(ctx context.Context, client *github.C
 	return repos, nil
 }
 
+func (s *Server) pushToGit(ctx context.Context, copyData func(dir, projPath string) error, repo, branch, subpath, token string, force bool) error {
+	ctx, cancel := context.WithTimeout(ctx, archivePullTimeout)
+	defer cancel()
+
+	// generate a temp dir to extract the archive
+	dir, err := os.MkdirTemp(os.TempDir(), "extracted_archives")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// use a subfolder for working with git
+	gitPath := filepath.Join(dir, "proj")
+	// projPath is the target for extracting the archive
+	projPath := gitPath
+	if subpath != "" {
+		projPath = filepath.Join(projPath, subpath)
+	}
+	err = os.MkdirAll(projPath, fs.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	gitAuth := &githttp.BasicAuth{Username: "x-access-token", Password: token}
+
+	var ghRepo *git.Repository
+	empty := false
+	ghRepo, err = git.PlainClone(gitPath, false, &git.CloneOptions{
+		URL:           repo,
+		Auth:          gitAuth,
+		ReferenceName: plumbing.NewBranchReferenceName(branch),
+		SingleBranch:  true,
+	})
+	if err != nil {
+		if errors.Is(err, transport.ErrEmptyRemoteRepository) {
+			empty = true
+			ghRepo, err = git.PlainInitWithOptions(gitPath, &git.PlainInitOptions{
+				InitOptions: git.InitOptions{
+					DefaultBranch: plumbing.NewBranchReferenceName(branch),
+				},
+				Bare: false,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to init git repo: %w", err)
+			}
+		} else {
+			return fmt.Errorf("failed to init git repo: %w", err)
+		}
+	}
+
+	wt, err := ghRepo.Worktree()
+	if err != nil {
+		return fmt.Errorf("failed to get worktree: %w", err)
+	}
+
+	var wtc worktreeContents
+	if !empty {
+		wtc, err = readWorktree(wt, subpath)
+		if err != nil {
+			return fmt.Errorf("failed to read worktree: %w", err)
+		}
+
+		if len(wtc.otherPaths) > 0 && !force {
+			return fmt.Errorf("worktree has additional contents")
+		}
+	}
+
+	// remove all the other paths
+	for _, path := range wtc.otherPaths {
+		err = os.RemoveAll(filepath.Join(projPath, path))
+		if err != nil {
+			return err
+		}
+	}
+
+	err = copyData(dir, projPath)
+	if err != nil {
+		return fmt.Errorf("failed to copy data: %w", err)
+	}
+
+	// add back the older gitignore contents if present
```

Comment:
If we are anyways overwriting all other files then I think it should be okay to also overwrite `.gitignore` file ?

Reply:
This was a requirement to not completely replace the existing file. But rather append like we do with `rill init`

Reply:
IIRC in `rill init` we did not delete existing files so it made sense to not remove existing `.gitignore`.
Here we are deleting all existing files as well.

Reply:
@nishantmonu51 @ericokuma  Thoughts on the above?